### ISSUE TITLE
SPARQL queries that return Integer 0 fail in AllegroGraph

### ISSUE THREAD

**Issue by mdorf** [ncbo/goo#104]
GROUP BY COUNT (or any other Integer result) queries that yield 0 rows fail in AllegroGraph

**Comment by mdorf** [comment]
Integer 0 results return format from each data store:
```
AllegroGraph: "{\"head\":{\"vars\":[\"g\", \"c\"]},\"results\":{\"bindings\":[\n {\"c\":{\"type\":\"literal\",\"datatype\":\"http://www.w3.org/2001/XMLSchema#integer\", \"value\":\"0\"}}]}}"
4store: "{\"head\":{\"vars\":[\"g\",\"c\"]},\n \"results\": {\n  \"bindings\":[]\n }}\n"
```

**Comment by mdorf** [comment]
Implemented a somewhat ugly fix that intercepts the results and sets the bindings to [] if the AllegroGraph format is encountered. May need to revisit this at a later time.

**Comment by mdorf** [comment]
This problem is caused by the inconsistent result from 4store. When the query returns 0, the format of the return is different than when the query returns non-zero:

4store:
-------
```
=> [] # this is a zero result
=> [#<RDF::Query::Solution:0x3ff25d4ff51c({:c=>#<RDF::Literal::Integer:0x3ff25f044444("33"^^<http://www.w3.org/2001/XMLSchema#integer>)>})>] # this is a non-zero result
```

AG:
----
```
=> [#<RDF::Query::Solution:0x3fd540aae7e0({:c=>#<RDF::Literal::Integer:0x3fd540aae86c("0"^^<http://www.w3.org/2001/XMLSchema#integer>)>}) >] # this is a zero result
=> [#<RDF::Query::Solution:0x3ff25d4ff51c({:c=>#<RDF::Literal::Integer:0x3ff25f044444("33"^^<http://www.w3.org/2001/XMLSchema#integer>)>})>] # this is a non-zero result
```

**Comment by mdorf** [comment]
Revision on the previous comment after additional testing:

The AG appears to return results even when the graph is null:

4store
------
```
[#<RDF::Query::Solution:0x3fc716b97b44({:g=>#<RDF::URI:0x3fc716b97dc4(http://data.bioontology.org/ontologies/MAPPING_TEST2/submissions/22)>, :c=>#<RDF::Literal::Integer:0x3fc716b97ce8("2"^^<http://www.w3.org/2001/XMLSchema#integer>)>})>]
```

AG
---
```
[#<RDF::Query::Solution:0x3fc8a9e69234({:c=>#<RDF::Literal::Integer:0x3fc8a9e692fc("0"^^<http://www.w3.org/2001/XMLSchema#integer>)>})>]
```

In the latter output, the entry for `:g` is missing.

**Comment by mdorf** [comment]
This appears to be a general difference between 4store and AllegroGraph. 4store ALWAYS matches the variables "vars" with bindings "bindings", even if the binding for a given var is empty. For example:
```
{
  "head": {
    "vars": ["id", "classType"]
  },
  "results": {
    "bindings":[
      {
        "id": {"type":"uri","value":"http://data.bioontology.org/ontologies/MAPPING_TEST4/submissions/44"},
        "classType": {}
      }
    ]
  }
}
```
The corresponding records from AG:
```
{
  "head": {
    "vars": ["id", "classType"]
  },
  "results": {
    "bindings": [
      {
        "id": {"type":"uri","value":"http://data.bioontology.org/ontologies/MAPPING_TEST2/submissions/22"}
      }
    ]
  }
}

This is problematic because one has to always provide additional checks on whether the binding exists, even though there is a corresponding variable for it.

**Comment by mdorf** [comment]
There appear to be a number of queries during the run of the Mappings test suite that return a single result in **AllegroGraph** that consists of the NULL value for the Graph and a "0" value for the corresponding COUNT. The equivalent queries in **4store** return NO RESULTS. Here is a screenshot of the **AllegroGraph** response:
<img width="722" alt="Screen Shot 2020-04-30 at 6 52 34 PM" src="https://user-images.githubusercontent.com/2042070/80775194-cb51da00-8b13-11ea-87c6-3f66464e3d47.png">

This produces the following response:

```
{
  "head": {
    "vars": ["g", "c"]
  },
  "results": {
    "bindings": [
      {
        "c": {"type":"literal","datatype":"http://www.w3.org/2001/XMLSchema#integer", "value":"0"}
      }
    ]
  }
}
```

The equivalent query in **4store** returns NO RESULTS and the following response:
```
{
  "head": {
    "vars": ["g", "c"]
  },
  "results": {
    "bindings": []
  }
}
```

**Comment by mdorf** [comment]
The queries that behave differently in **AllegroGraph** vs **4store**:

```
SELECT ?g (count(?s1) as ?c)
WHERE {
  {
    GRAPH <http://data.bioontology.org/ontologies/MAPPING_TEST4/submissions/44> {
      ?s1 <http://bioportal.bioontology.org/ontologies/umls/cui> ?o .
    }
    GRAPH ?g {
      ?s2 <http://bioportal.bioontology.org/ontologies/umls/cui> ?o .
    }
  }
  FILTER (?s1 != ?s2)
  FILTER (!STRSTARTS(str(?g),'http://data.bioontology.org/ontologies/MAPPING_TEST4'))
} GROUP BY ?g
```
```
SELECT ?g (count(?s1) as ?c)
WHERE {
  {
    GRAPH <http://data.bioontology.org/ontologies/MAPPING_TEST4/submissions/44> {
        ?s1 <http://data.bioontology.org/metadata/def/mappingSameURI> ?o .
    }
    GRAPH ?g {
        ?s2 <http://data.bioontology.org/metadata/def/mappingSameURI> ?o .
    }
  }
  FILTER (!STRSTARTS(str(?g),'http://data.bioontology.org/ontologies/MAPPING_TEST4'))
} GROUP BY ?g
```
```
SELECT ?g (count(?s1) as ?c)
WHERE {
  {
    GRAPH <http://data.bioontology.org/ontologies/MAPPING_TEST4/submissions/44> {
      ?s1 <http://data.bioontology.org/metadata/def/mappingLoom> ?o .
    }
    GRAPH ?g {
      ?s2 <http://data.bioontology.org/metadata/def/mappingLoom> ?o .
    }
  }
  FILTER (?s1 != ?s2)
  FILTER (!STRSTARTS(str(?g),'http://data.bioontology.org/ontologies/MAPPING_TEST4'))
} GROUP BY ?g
```
```
SELECT ?g (count(?s1) as ?c)
WHERE {
  {
    GRAPH <http://data.bioontology.org/ontologies/MAPPING_TEST4/submissions/44> {
      ?s1 <http://data.bioontology.org/metadata/def/mappingRest> ?o .
    }
    GRAPH ?g {
      ?s2 <http://data.bioontology.org/metadata/def/mappingRest> ?o .
    }
  }
  FILTER (?s1 != ?s2)
  FILTER (!STRSTARTS(str(?g),'http://data.bioontology.org/ontologies/MAPPING_TEST4'))
} GROUP BY ?g
```

**Comment by mdorf** [comment]
The issue is still valid in the AllegroGraph v7.0.0
<img width="734" alt="Screen Shot 2020-05-04 at 4 18 36 PM" src="https://user-images.githubusercontent.com/2042070/81022710-28ef6a80-8e23-11ea-8132-ea24dacb8b65.png">

**Comment by mdorf** [comment]
Further observed that the same single result is returned by the AllegroGraph even when a bogus graph names are used:

<img width="778" alt="Screen Shot 2020-05-04 at 4 40 26 PM" src="https://user-images.githubusercontent.com/2042070/81023688-014dd180-8e26-11ea-81c1-8cb37b059afa.png">

**Comment by mdorf** [comment]
Here is a response from AllegroGraph support on this:

> Regarding issue 2, the query can be simplified to:
> 
>  select ?s (count(?o) AS ?count)
>  where {
>    ?s <ex:p> ?o
>  } group by ?s
> 
> and if there are no matching triples, the result in AllegroGraph is one row with ?s unbound and ?count 0, whereas you might expect zero rows.
> 
> It turns out the SPARQL 1.1 standard is ambiguous, making either of them acceptable, and it is up to the implementation. However I found that the consensus now is that the desired output is zero rows, similar to SQL. For example Andy Seaborne, who edited the SPARQL 1.1 specification, has written about exactly this:
> https://afs.github.io/sparql-agg-group-empty.html#group-agg-no-rows
> 
> **We aim to follow the SPARQL 1.1 standard in text and spirit, so we'll follow the consensus and modify AllegroGraph's behaviour.**
> 
> Thanks for bringing this to our attention.

**Comment by mdorf** [comment]
Another response from AllegroGraph support:

> **We plan to do this in the next two weeks.**
> Meanwhile you could probably rewrite the query by introducing outer select and filter, like:
> 
> select * {
>   {
>     select ?s (count(?o) as ?c) {
>       ?s <ex:p> ?o
>     } group by ?s
>   }
>   filter (bound(?s))
> }
>

**Comment by mdorf** [comment]
This appears to be fixed in AllegroGraph v7.0.1

**Comment by mdorf** [comment]
The code that handles this case has been removed.
